File: GGAndroid_2.0/GGControls/GGViewModel/src/main/java/io/ganguo/viewmodel/common/base/BaseHFRViewModel.java
package io.ganguo.viewmodel.common.base;

import android.databinding.ObservableBoolean;
import android.databinding.ViewDataBinding;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;

import io.ganguo.library.ui.adapter.v7.LoadMoreListener;
import io.ganguo.library.ui.view.ViewInterface;
import io.ganguo.utils.util.Views;
import io.ganguo.utils.util.log.Logger;
import io.ganguo.viewmodel.callback.ICenterLoadingView;
import io.ganguo.viewmodel.common.RecyclerViewModel;
import io.ganguo.vmodel.BaseViewModel;
import io.ganguo.vmodel.VMLifecycleHelper;
import io.ganguo.vmodel.ViewModelHelper;
import io.ganguo.vmodel.adapter.ViewModelAdapter;

/**
 * <p>
 * HFRecyclerViewModel && HFSRecyclerViewModel 基类
 * </p>
 * Created by leo on 2018/8/21.
 */
public abstract class BaseHFRViewModel<B extends ViewDataBinding, T extends ViewInterface<B>> extends BaseViewModel<T>
        implements LoadMoreListener, ViewModelAdapter.OnNotifyDiffUtilDataChangedListener {
    private RecyclerViewModel<BaseViewModel, ViewDataBinding> recyclerViewModel;
    public ObservableBoolean isAnimateLayoutChanges = new ObservableBoolean(true);
    public ObservableBoolean isEnableHeaderElevation = new ObservableBoolean(true);
    public ObservableBoolean isEnableFooterElevation = new ObservableBoolean(true);
    protected ICenterLoadingView centerLoadingView;


    @Override
    public void onAttach() {
        super.onAttach();
        initHeader(getHeaderContainer());
        initFooter(getFooterContainer());
        initEmpty(getEmptyContainer());
        initRecyclerView();
        initLoadingView();
    }


    /**
     * function:配置加载Loading
     * 注意：如果设置了skeletonScreen，则优先使用skeletonScreen！
     */
    public void initLoadingView() {
        initCenterLoading();
        showCenterLoading();
    }


    /**
     * init center Loading
     */
    protected void initCenterLoading() {
        centerLoadingView = createLoadingView();
        if (centerLoadingView == null) {
            return;
        }
        ViewModelHelper.bind(getLoadingContainer(), this, centerLoadingView.getViewModel());
    }


    /**
     * init RecyclerView
     */
    protected void initRecyclerView() {
        ViewModelHelper.bind(getRecyclerRootView(), this, recyclerViewModel = initRecyclerViewModel());
        getAdapter().setLoadMoreListener(this);
        //set recyclerView default match parent
        initRecyclerViewHeight();
        initRecyclerViewAnimator();
        getAdapter().setDataChangedListener(this);
    }


    /**
     * init RecyclerView ItemAnimator
     */
    protected void initRecyclerViewAnimator() {
        RecyclerView.ItemAnimator itemAnimator = getRecyclerView().getItemAnimator();
        int duration = 300;
        itemAnimator.setAddDuration(duration);
        itemAnimator.setChangeDuration(duration);
        itemAnimator.setMoveDuration(duration);
        itemAnimator.setRemoveDuration(duration);
    }

    /**
     * 创建RecyclerViewViewModel
     *
     * @return
     */
    protected RecyclerViewModel<BaseViewModel, ViewDataBinding> initRecyclerViewModel() {
        return RecyclerViewModel
                .linerLayout(getContext(), LinearLayoutManager.VERTICAL)
                .setViewHeight(ViewGroup.LayoutParams.MATCH_PARENT);
    }


    /**
     * 设置RecyclerView高度是否MatchParent
     *
     * @return
     */
    protected void initRecyclerViewHeight() {
        ViewGroup.LayoutParams params = getRecyclerView().getLayoutParams();
        params.height = ViewGroup.LayoutParams.MATCH_PARENT;
        getRecyclerView().setLayoutParams(params);
    }

    /**
     * 隐藏Loading并判断是否显示EmptyView
     */
    public void toggleEmptyView() {
        if (!isAttach()) {
            Logger.d("call: " + VMLifecycleHelper.EVENT_VIEW_MODEL_DETACH);
            return;
        }
        hideCenterLoading();
        if (getAdapter().size() == 0) {
            Views.visible(getEmptyContainer());
        } else {
            Views.gone(getEmptyContainer());
        }
    }

    /**
     * 显示中心Loading
     */
    public void showCenterLoading() {
        Views.visible(getLoadingContainer());
        if (centerLoadingView != null) {
            centerLoadingView.onStartLoading();
        }
    }

    /**
     * 隐藏中心Loading
     */
    public void hideCenterLoading() {
        Views.gone(getLoadingContainer());
        if (centerLoadingView != null) {
            centerLoadingView.onStopLoading();
        }
    }


    public void setAnimateLayoutChanges(boolean isAnimate) {
        isAnimateLayoutChanges.set(isAnimate);
    }

    public void setEnableFooterElevation(boolean isEnable) {
        isEnableFooterElevation.set(isEnable);
    }

    public void setEnableHeaderElevation(boolean isEnable) {
        isEnableHeaderElevation.set(isEnable);
    }


    /**
     * Getters
     */
    public RecyclerViewModel<BaseViewModel, ViewDataBinding> getRecyclerViewModel() {
        return recyclerViewModel;
    }

    public RecyclerView getRecyclerView() {
        if (recyclerViewModel == null) {
            return null;
        }
        return recyclerViewModel.getRecyclerView();
    }

    public ViewModelAdapter<ViewDataBinding> getAdapter() {
        if (recyclerViewModel == null) {
            return null;
        }
        return recyclerViewModel.getAdapter();
    }

    public abstract ViewGroup getHeaderContainer();

    public abstract ViewGroup getFooterContainer();

    public abstract ViewGroup getEmptyContainer();

    public abstract ViewGroup getLoadingContainer();

    public abstract ViewDataBinding getRecyclerRootView();

    protected abstract ICenterLoadingView createLoadingView();


    /**
     * Bind ViewModel to container here
     *
     * @see {@link ViewModelHelper#bind(ViewGroup, BaseViewModel, int)}
     */
    public void initHeader(ViewGroup container) {
    }

    public void initFooter(ViewGroup container) {
    }

    public void initEmpty(ViewGroup container) {

    }

    @Override
    public void onLoadMore() {
        //进入LoadMore函数后，强制停止RecyclerView滑动，避免滑动过快，出现多次LoadMore的情况！
        Views.onCancelRecyclerViewScroll(getRecyclerView());
    }

    @Override
    public void onAdapterNotifyComplete() {

    }
}

File: GGAndroid_3.0/core/frame/mvvm-v1/src/main/java/io/ganguo/viewmodel/pack/common/base/BaseHFRViewModel.kt
package io.ganguo.viewmodel.pack.common.base

import android.view.ViewGroup
import androidx.databinding.ObservableBoolean
import androidx.databinding.ViewDataBinding
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import io.ganguo.viewmodel.R
import io.ganguo.viewmodel.core.ViewModelHelper
import io.ganguo.viewmodel.core.adapter.ViewModelAdapter
import io.ganguo.viewmodel.core.adapter.diffuitl.ViewModelDiffUtil
import io.ganguo.viewmodel.core.viewinterface.ViewInterface
import io.ganguo.viewmodel.databinding.IncludeRecyclerBinding
import io.ganguo.viewmodel.pack.base.viewmodel.BaseSupportStateViewModel
import io.ganguo.viewmodel.pack.common.RecyclerViewModel

/**
 *
 *
 * HFRecyclerViewModel && HFSRecyclerViewModel 基类
 *
 * Created by leo on 2018/8/21.
 */
abstract class BaseHFRViewModel<B : ViewDataBinding, T : ViewInterface<B>> : BaseSupportStateViewModel<T>() {
    override val stateLayout: ViewGroup by lazy {
        statusContainer
    }

    /**
     * Getters RecyclerViewModel
     *
     * @return
     */
    private val recyclerVModel: RecyclerViewModel<IncludeRecyclerBinding, ViewInterface<IncludeRecyclerBinding>> by lazy {
        initRecyclerViewModel()
    }

    /**
     * Getters RecyclerView
     *
     * @return
     */
    val recyclerView: RecyclerView by lazy {
        recyclerVModel.recyclerView
    }

    /**
     * Getters Adapter
     *
     * @return
     */
    val adapter: ViewModelAdapter by lazy {
        recyclerVModel.adapter
    }
    var isAnimateLayoutChanges = ObservableBoolean(true)
    var isEnableHeaderElevation = ObservableBoolean(false)
    var isEnableFooterElevation = ObservableBoolean(false)


    /**
     * 页面背景颜色
     *
     * @return
     */
    open val backgroundColorRes: Int by lazy {
        R.color.transparent
    }

    /**
     * getters abstract ViewGroup
     *
     * @return
     */
    abstract val headerContainer: ViewGroup
    abstract val footerContainer: ViewGroup
    abstract val statusContainer: ViewGroup
    abstract val recyclerRootView: ViewDataBinding

    override fun onAttach() {
        super.onAttach()
        rootView.setBackgroundResource(backgroundColorRes)
        initHeader(headerContainer)
        initFooter(footerContainer)
        initRecyclerView()
    }

    /**
     * init RecyclerView
     */
    protected open fun initRecyclerView() {
        val adapter = ViewModelAdapter(context, this, ViewModelDiffUtil())
        recyclerVModel.setRecyclerAdapter(adapter)
        ViewModelHelper.bind(recyclerRootView, this, recyclerVModel)
        initRecyclerViewAnimator()
    }

    /**
     * 创建RecyclerViewViewModel
     */
    protected open fun initRecyclerViewModel(): RecyclerViewModel<IncludeRecyclerBinding, ViewInterface<IncludeRecyclerBinding>> {
        return RecyclerViewModel
                .linerLayout<IncludeRecyclerBinding, ViewInterface<IncludeRecyclerBinding>> (context, LinearLayoutManager.VERTICAL)
                .isOverScroll(false)
                .setViewHeight(ViewGroup.LayoutParams.MATCH_PARENT)
    }

    /**
     * init RecyclerView ItemAnimator
     */
    protected open fun initRecyclerViewAnimator() {
        val itemAnimator = recyclerView.itemAnimator
        val duration = 300
        itemAnimator!!.addDuration = duration.toLong()
        itemAnimator.changeDuration = duration.toLong()
        itemAnimator.moveDuration = duration.toLong()
        itemAnimator.removeDuration = duration.toLong()
    }

    /**
     * 隐藏Loading并判断是否显示EmptyView
     */
    open fun toggleEmptyView() {
        if (!isAttach) {
            return
        }
        if (adapter.itemCount == 0) {
            showEmptyView()
        } else {
            showContentView()
        }
    }

    /**
     * 是否开启View状态改变动画
     *
     * @param isAnimate 是否开启
     */
    fun setAnimateLayoutChanges(isAnimate: Boolean) {
        isAnimateLayoutChanges.set(isAnimate)
    }

    /**
     * 是否开启Footer底部阴影
     *
     * @param isEnable 是否开启
     */
    fun setEnableFooterElevation(isEnable: Boolean) {
        isEnableFooterElevation.set(isEnable)
    }

    /**
     * 是否开启header顶部阴影
     *
     * @param isEnable 是否开启
     */
    fun setEnableHeaderElevation(isEnable: Boolean) {
        isEnableHeaderElevation.set(isEnable)
    }


    /**
     * 初始化顶部菜单布局
     *
     * @param container
     */
    open fun initHeader(container: ViewGroup) {}

    /**
     * 初始化底部菜单布局
     *
     * @param container
     */
    open fun initFooter(container: ViewGroup) {}
}